### ISSUE TITLE
[Bug]: dock alignment doesn't work after setting initial state to withdrawn

### ISSUE THREAD

**Issue by Bluey26** [brndnmtthws/conky#2112]
### What happened?

For conky versions newer than `1.21.7` there seems to be a X11 regression in where the conky window is not showing itself anymore in the given position. `alignment = 'top_right'` seems to be ignored, and the window is shown in the top_left border, transparency is not detected anymore (`own_window_argb_value = 90` ignored) and there is also a "windowing" issue, since the `own_window_type = 'dock'` property is also ignored (the conky panel "reserves" the space and avoids other windows and the desktop itself to be on top of it.

I discovered this after compiling the lastest changes in conky, but it seems to be a problem arisen from prior changes. Downgrading my conky to `conky-git-1.21.7.r5.g71814776-1` seems to solve the issue, for some reason (x11 backend now has all the mentioned properties working).

This seems to be independent of the wayland_support argument. For some reason, the issue is not shown when using the **wayland backend** in a Wayland session (i.e. labwc)

**PS**: I don't think that my conky config is relevant in this case, because the .conf is the same for the wayland backend(1.21.9) and for the previous version (1.21.7), in both cases giving me a working conky.

**PS2:** Since there is a WIP wayland_backend that works now, i was thinking if it would be possible to have a conky property that lets the user to have a "preference" over the backend use, something like : `prefer_backend = wayland` or `prefer_wayland = 1`. This is an issue because if somebody uses x11 and wayland from time to time,  `out_to_wayland = true, out_to_x = true,` will give the x11 (xwayland) backend in wayland instead of the newer wayland backend.

### Version

Lastest (1.21.9)

### Which OS/distro are you seeing the problem on?

Arch Linux

**Comment by Caellian** [comment]
Related PR in 1.21.8 is #2056.
But I believe that's not the cause of the issue and only unearths some other problem.
Can you check whether you're experiencing issue described in #2046 on 1.21.7?

Commits that touch `set_struts`:
- c5ee70177c7d9b586707cf1adefdc45894c5cdea
  - first change by me that could cause issues; set_struts rewrite
- 7b44b2a27bd0649c7c00a4f5f4ec521aa554777b
  - fixes missing variables from previous change
- 6bf383cf786a9615480b0268571cb2ad84e2fd24
  - introduces geometry primitives which get used in struts
- 2d50767216d2555b2d8d62765abaf41cb6d6979a (last in 1.21.7)
  - adds utilities to geometry primitives
- 9f830224e59ac9053415ffae1c363ad9020fc783
  - fixes commented out code as described in #2056.

**Comment by Bluey26** [comment]
> Can you check whether you're experiencing issue described in #2046 on 1.21.7?

If i understand the issue correctly, i should see the conky dock window over all the windows, in "top" of them.

That does not happen in my build,fortunately.

I was looking at the compiled package i have for conky 1.21.7 and it was compiled on 2024-09-17 (17th Sept), thus it should be the same as : https://github.com/brndnmtthws/conky/tree/71814776d9e54033a7a5fe8427c43504ea8f1f8d

This means my package is not the 1.21.7 release, but a middle between 1.21.7 and 1.21.8.

The change that may be triggering my issue is somewhat recent, something after 17th september.

I will compile that commit again and see what happens with the compiled package, just to discard some kind of compilation issue(newer dependencies that may cause the issue mentioned in here).

**Comment by Bluey26** [comment]
I compiled a intermediate commit (https://github.com/brndnmtthws/conky/commit/ebef88015ae00efe30553ae68d45154985598e5d) which seems to have a regression regards the issue. This commit has the position issue, and seems likely to be the one invoved ("fix incorrect x11 dock/panel placement).

This has introduced a regression, at least in here.

**Comment by Caellian** [comment]
Right, but that change [was as is before I touched that code](https://github.com/brndnmtthws/conky/blob/c51fbe66f63b365dcf4918a6d282c4a6af42225d/src/x11.cc#L737), which is why I said ebef88015ae00efe30553ae68d45154985598e5d should be ignored. I commented it out as part of 6adf6b9dd4d368640bf7ef57f3e6199d83154d78 ([here](https://github.com/brndnmtthws/conky/commit/6adf6b9dd4d368640bf7ef57f3e6199d83154d78#diff-c9ac161ed259f39c83ac37ec6b483503331c4c57d081322d1c373c9ff938b9d6L735-R611)), by accident.

What's worth comparing is the `set_struts` implementation and changes in used variables. 

<details>
<summary><h3>Original implementation</h3></summary>

```c++
/* reserve window manager space */
void set_struts(int sidenum) {
  Atom strut;
  if ((strut = ATOM(_NET_WM_STRUT)) != None) {
    /* reserve space at left, right, top, bottom */
    signed long sizes[12] = {0};
    int i;

    /* define strut depth */
    switch (sidenum) {
      case 0:
        /* left side */
        sizes[0] = window.x + window.width;
        break;
      case 1:
        /* right side */
        sizes[1] = display_width - window.x;
        break;
      case 2:
        /* top side */
        sizes[2] = window.y + window.height;
        break;
      case 3:
        /* bottom side */
        sizes[3] = display_height - window.y;
        break;
    }

    /* define partial strut length */
    if (sidenum <= 1) {
      sizes[4 + (sidenum * 2)] = window.y;
      sizes[5 + (sidenum * 2)] = window.y + window.height;
    } else if (sidenum <= 3) {
      sizes[4 + (sidenum * 2)] = window.x;
      sizes[5 + (sidenum * 2)] = window.x + window.width;
    }

    /* check constraints */
    for (i = 0; i < 12; i++) {
      if (sizes[i] < 0) {
        sizes[i] = 0;
      } else {
        if (i <= 1 || i >= 8) {
          if (sizes[i] > display_width) { sizes[i] = display_width; }
        } else {
          if (sizes[i] > display_height) { sizes[i] = display_height; }
        }
      }
    }

    XChangeProperty(display, window.window, strut, XA_CARDINAL, 32,
                    PropModeReplace, reinterpret_cast<unsigned char *>(&sizes),
                    4);

    if ((strut = ATOM(_NET_WM_STRUT_PARTIAL)) != None) {
      XChangeProperty(display, window.window, strut, XA_CARDINAL, 32,
                      PropModeReplace,
                      reinterpret_cast<unsigned char *>(&sizes), 12);
    }
  }
}
```

</details>

I need to write a test for it and figure out what I broke.

**Comment by Caellian** [comment]
I'm working on adding tests for `set_struts` in #2115.

**Comment by Caellian** [comment]
Which WM are you on?

**Comment by Bluey26** [comment]
>Which WM are you on?
I use labwc, which is based in wlroots

**Comment by Caellian** [comment]
That's odd, the current `set_struts` implementation _should_ work with openbox because it adheres to that standard. I'll probably need to compare our code with what `tint2` does and hope we don't have to switch between #2056 and commented out version based on WM.

I tested this code on openbox and commented out version worked correctly, it's possible that the uncommented one didn't which is why I could've left it commented out. Just X11 things I guess... hahah

There _should_ exist a version that works for both without having to check for WM.

`fluxbox` doesn't adhere to the spec with regards to struts and #2056 fixes this code for that WM (and probably compiz, kwin, i3 as well).

**Comment by Bluey26** [comment]
In my opinion it would be useful to have such option in the meanwhile. Moreover, if the releases do not have those, there may be issues with stable Linux distributions which still are using Openbox as WM.

**Comment by Caellian** [comment]
Can you give a try to [`fix/struts`](https://github.com/brndnmtthws/conky/tree/fix/struts) branch and let me know how the windows behave with new changes?

**Comment by Bluey26** [comment]
Hello, the issue persists in here. I compiled the branch using:

```
source=("git+https://github.com/brndnmtthws/conky#branch=fix/struts"
		"$pkgname-use-system-cmake.patch")
```
(see https://gitlab.archlinux.org/archlinux/packaging/packages/conky/-/blob/main/PKGBUILD?ref_type=heads)

I logged into X11 and the panel is shown over the left border(i set it to be in the top-right), it reserves the space of the desktop(this is not a "dock" behavior in Wayland, and it was not normal in previous versions), it does not respects the color (transparency,picom running) and it is set in a way that avoids the windows or the icons to be over it.

To show how serious the issue is, i recorded this small video:

https://github.com/user-attachments/assets/e0c28b46-2646-47dd-8022-4a5cb303d471

You can see that the conky panel is not honoring any of the given properties i was describing in the previous comments, and that it is not possible to maximize the firefox window over it. I have hidden my desktop items, but they are not able to be set into the bottom of the conky panel, as it is occupying all the desktop (from the top of the conky window up to my panel).

None of this is seen in wayland,for some reason.

**Comment by Caellian** [comment]
Wayland completely ignores how panels are supposed to work currently. So we can ignore it for this issue - there's a separate one for that. Also, there's a separate issue for transparency. I'll link them later when I'll be at a PC. Let's keep this issue focused on space reserving functionality only.

Are you sure you're using a `dock` window and not `panel`? We don't set struts unless the window is `panel`.

I tested things out on Openbox, and conky is positioned properly with `panel`.

Do you mind providing your config so I can try it out?

**Comment by Caellian** [comment]
Yeah, `dock` doesn't reserve space. Panel and dock are pretty much the same, conky disambiguates the two though:
- both are placed below widnows and above desktop,
- `dock` doesn't reserve screen space,
- `panel` does.

**Comment by Bluey26** [comment]
The problem is the contrary: i have set dock because i don't want it to reserve space, but the compiled branch fix/struts seems to be doing it anyway, and making unable to place windows over the conky _dock_

**Comment by Caellian** [comment]
Hmm.... let me check...

As for reserved space - it will always reserve entire space along the edge. Openbox workarea is a rectangle and all docks/panels can only make it smaller in 4 directions; i.e. we can't cut out a corner for conky because Openbox doesn't support that.

**Comment by Bluey26** [comment]
Hello again, i saw the notification and compiled again. Unfortunately this does not solve the issue for me, and changing to `desktop` or `normal` gives me other issues in here (unfortunately dock is the only possible for me).

So i decided to make a patch so i can use the latest git branch without issues. I know that this commit (https://github.com/brndnmtthws/conky/commit/ebef88015ae00efe30553ae68d45154985598e5d.patch) is the one gives me troubles, so i wanted to revert it.

I am struggling to do it, by chance could you help me to do the correct .patch file?

The code itself would be the opposite:

```
 #endif /* BUILD_XSHAPE */
       if (own_window_type.get(l) == window_type::DOCK ||
           own_window_type.get(l) == window_type::PANEL) {

-        // Docks and panels MUST have WithdrawnState initially
-        // See: https://github.com/brndnmtthws/conky/issues/2046
-        wmHint.initial_state = WithdrawnState;
+        // wmHint.initial_state = WithdrawnState;
       } else {
         wmHint.initial_state = NormalState;
       }
```

But i am unable to apply the patch using archlinux ABS system, because of the following:

`File to patch:`

Which seems to be an issue with the previous section of the patch(defining lines?):

```
diff --git a/src/x11.cc b/src/x11.cc
index f44d5d52d..db6c7a5ac 100644
--- a/src/x11.cc
+++ b/src/x11.cc
@@ -640,7 +640,9 @@
```

I tried to change those to point to the x11.cc file without luck

**Comment by Caellian** [comment]
What kind of issues are you facing with `normal`/`desktop` window?

Looking into the past commits, `initial_state` was [always](https://github.com/brndnmtthws/conky/commit/92bdce5d1a17a6e85f047efb049ceaafbc58a3c2#diff-ee06dee9a439cba1cb913824c453a666160f258c9adf933b539b9d8d5b93b44cR270-R271) `WithdrawnState`, and it's actually necessary to do that on some WMs before mapping the window to make the WMs properly read out window hints.

The behavior where `dock` reserves space even though we only set them for `panel` is due to how Openbox handles the window. I tried setting struts to 0, but that didn't make Openbox behave any different.

**Comment by Bluey26** [comment]
I think this is quite a niche issue, but here is my issue with normal/desktop modes:

I use a DE called LXQt, which has a menu plugin called "show desktop". Whenever i use that plugin to minimize all the windows, Conky disappears of the screen. Conky is somewhat taken as a window and thus is minimized. Fortunately this does not happen with the `dock` option, so i have been using that one for several years, as long as i can recall.

This happens in Openbox, and i think it can't be helped. I also remember that there was an issue with normal or desktop mode, in where conky disappeared of the screen when i selected items on it.

**Comment by Caellian** [comment]
Alright, that's a completely different issue then. `normal`/`desktop` should work for your use case, we're just not setting `_NET_WM_ALLOWED_ACTIONS` to prevent conky from being minimized.

**Comment by Bluey26** [comment]
Yet it is not clear to me why i should use `normal `or `desktop`.` dock` has been working all this time for me. 

Without the `sed` line, in wayland (labwc,wlroots based) the dock panel shows properly as it should, its just that line that messes Openbox behavior and does not respect all the options for me. 

Even if the dock reserves all the space, it should honor all the other properties: color (transparent), position(top right), which has not been the case as you can see in the X11 video i shared previously.

I still think there is something wrong in the openbox code, yet i am no programmer or know the conky code, its just a feeling i have.

**Comment by Caellian** [comment]
I did some more testing on Openbox and you're right. `xprop` doesn't show _any_ properties being set for conky window if it's started in `WithdrawnState`. I created a new PR that makes the window start in `WithdrawnState` only for Fluxbox which handles it differently.

Spec/docs don't really specify when `WithdrawnState` should be used, but yeah.

As for the transparency thing, please refer to #2015 if the last PR doesn't fix that problem as well.

**Comment by Bluey26** [comment]
I think the transparency issue is also related to what you describe. The older version of conky i have, shows successfully the transparency, with the given options i've shared in a previous comment. When the `WithdrawnState` line is commented, the transparency also works(picom is loaded on `startup`) in the latest version,compiled with the sed line.

**Comment by Bluey26** [comment]
**Update:** The new changes done into master branch (Withdrawn state only for fluxbox) seems to solve the issue, now my conky dock looks exactly as the older versions and the wayland one (transparency too). Thanks!
